Require explicit counter_cache option for has_many

tristang opened the pull request:
Previously has_many associations assumed a counter_cache was to be used
based on the presence of an appropriately named column. This is
inconsistent, since the inverse belongs_to association will not make
this assumption. See issues #19042 #8446.
This commit checks for the presence of the counter_cache key in the
options of either the has_many or belongs_to association as well as
ensuring that the *_count column is present.

tristang committed:
Require explicit counter_cache option for has_many

Previously has_many associations assumed a counter_cache was to be used
based on the presence of an appropriately named column. This is
inconsistent, since the inverse belongs_to association will not make
this assumption. See issues #19042 #8446.
This commit checks for the presence of the counter_cache key in the
options of either the has_many or belongs_to association as well as
ensuring that the *_count column is present.

tristang commented:
Hey Rails team. This is my first PR, so let me know if I've made any mistakes in the process. Cheers.

rafaelfranca commented:
When you say:

> This is
> inconsistent, since the inverse belongs_to association will not make
> this assumption.

Means we have the same check at `belongs_to` side?

tristang commented:
That's right. `ActiveRecord::Reflection::AssociationReflection#counter_cache_column` returns `nil` unless the `options[:counter_cache]` option is given on the association explicitly. 

      def counter_cache_column
        if options[:counter_cache] == true
          "#{active_record.name.demodulize.underscore.pluralize}_count"
        elsif options[:counter_cache]
          options[:counter_cache].to_s
        end
      end

rafaelfranca commented:
Can we add a test that make sure the column is not used if we don't have the option?

tristang commented:
Sure. I'll add them today.

tristang committed:
Add tests for associations without counter_cache

Assert that counter_cache behaviour is not used on belongs_to or
has_many associations if the option is not given explicitly.

omnikron commented:
@rafaelfranca any chance of taking another look at this now that @tristang added the tests? I would love to see this fixed.

eduardohertz commented:
+1 for this fix too.

vfonic commented:
+1

t-cyrill commented:
+1

rafaelfranca commented:
Backported in e0e3715557caae111460598f3e78409578b8f393
